homeassistant/components/enasolar/config_flow.py

```diff
@@ -0,0 +1,204 @@
+"""EnaSolar solar inverter configuration."""
+from __future__ import annotations
+
+import logging
+from typing import Any
+
+import pyenasolar
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.core import HomeAssistant, callback
+import homeassistant.helpers.config_validation as cv
+from homeassistant.util import dt as dt_util
+
+from .const import (
+    CAPABILITY,
+    CONF_CAPABILITY,
+    CONF_DC_STRINGS,
+    CONF_HOST,
+    CONF_MAX_OUTPUT,
+    CONF_NAME,
+    CONF_SUN_DOWN,
+    CONF_SUN_UP,
+    DC_STRINGS,
+    DEFAULT_HOST,
+    DEFAULT_NAME,
+    DEFAULT_SUN_DOWN,
+    DEFAULT_SUN_UP,
+    DOMAIN,
+    MAX_OUTPUT,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@callback
+def enasolar_entries(hass: HomeAssistant):
+    """Return the hosts already configured."""
+    return {
+        entry.data[CONF_HOST] for entry in hass.config_entries.async_entries(DOMAIN)
+    }
+
+
+class EnaSolarConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """User Configuration of EnaSolar Integration."""
+
+    VERSION = 1
+
+    def __init__(self) -> None:
+        """Initialize the config flow."""
+        self._enasolar = pyenasolar.EnaSolar()
+        self._data: dict[str, Any] = {}
+        self._errors: dict[str, str] = {}
+
+    def _host_in_configuration_exists(self, host) -> bool:
+        """Return True if host exists in configuration."""
+        if host in enasolar_entries(self.hass):
+            return True
+        return False
+
+    async def _try_connect(self, host):
+        try:
+            await self._enasolar.interogate_inverter(host)
+            return True
+
+        except Exception as conerr:  # pylint: disable=broad-except
+            self._errors[CONF_HOST] = "cannot_connect"
+            _LOGGER.error(
+                "Connection to EnaSolar inverter '%s' failed (%s)", host, conerr
+            )
+        return False
+
+    def get_name(self):
+        """Needed to mock name when running tests."""
+        return self._data[CONF_NAME]
+
+    @staticmethod
+    @callback
+    def async_get_options_flow(config_entry):
+        """Get options flow for this handler."""
+        return EnaSolarOptionsFlowHandler(config_entry)
+
+    async def async_step_user(self, user_input=None):
+        """Step when user initializes a integration."""
+        _errors = {}
+
+        if user_input is not None:
+            self._data[CONF_HOST] = user_input[CONF_HOST]
+            if self._host_in_configuration_exists(self._data[CONF_HOST]):
+                return self.async_abort(reason="already_configured")
+            self._data[CONF_NAME] = user_input[CONF_NAME]
+            if await self._try_connect(self._data[CONF_HOST]):
+                await self.async_set_unique_id(self._enasolar.get_serial_no())
```

Comment:
Does this do I/O?

Reply:
_try_connect does a copy of HTTP requests.

Reply:
Does `self._enasolar.get_serial_no()` do any I/O?

Reply:
No, the value has already been retrieved by _try_connect

Reply:
Have moved this around a bit as I wanted to get the serial no earlier to detect duplicate configs

Reply:
Needed to add stub again to allow mocking out in the tests